src/v/cluster/node_status_provider.cc

```diff
@@ -0,0 +1,239 @@
+
+    if (node == _self) {
+        return node_status{
+          .node_id = node, .last_seen = rpc::clock_type::now()};
+    }
+
+    if (auto it = _peers_state.find(node); it != _peers_state.end()) {
+        return it->second;
+    }
+
+    return std::nullopt;
+}
+
+void node_status_provider::update_node_status(node_status_metadata meta) {
+    auto [it, inserted] = _discovered_peers.insert(meta.node_id);
+    if (inserted) {
+        vlog(
+          clusterlog.info, "Node {} has been discovered via RPC request", *it);
+    }
+
+    _peers_state[meta.node_id] = {
+      .node_id = meta.node_id, .last_seen = rpc::clock_type::now()};
+}
+
+void node_status_provider::tick() {
+    ssx::background = ssx::spawn_with_gate_then(_gate, [this] {
+                          return dispatch_node_status_requests().finally(
+                            [this] {
+                                if (!_gate.is_closed()) {
+                                    _timer.arm(_period());
+                                }
+                            });
+                      }).handle_exception([](const std::exception_ptr& e) {
+        vlog(
+          clusterlog.debug,
+          "Exception during node status provider tick: {}",
+          e);
+    });
+}
+
+ss::future<> node_status_provider::dispatch_node_status_requests() {
+    if (!_feature_table.local().is_active(feature::raftless_node_status)) {
+        return ss::make_ready_future<>();
+    }
+
+    node_status_request request = {.sender_metadata = {.node_id = _self}};
+
+    return ssx::parallel_transform(
+      _discovered_peers.begin(),
+      _discovered_peers.end(),
+      [this, request = std::move(request)](auto peer_id) {
+          return do_node_status(peer_id, request);
+      });
+}
+
+ss::future<> node_status_provider::do_node_status(
+  model::node_id target, node_status_request r) {
+    auto gate_holder = _gate.hold();
+
+    auto reply = co_await send_node_status_request(target, r);
+    if (reply.has_error()) {
+        auto err = reply.error();
+        if (
+          err.category() == rpc::error_category()
+          && static_cast<rpc::errc>(err.value())
+               == rpc::errc::client_request_timeout) {
+            _stats.rpcs_timed_out += 1;
+        }
+
+        vlog(
+          clusterlog.debug,
+          "Error occurred while sending node status request: {}",
+          err.message());
+
+        co_return;
+    }
+
+    _stats.rpcs_sent += 1;
+    process_reply(std::move(reply));
+}
+
+ss::future<result<node_status_reply>>
+node_status_provider::send_node_status_request(
+  model::node_id target, node_status_request r) {
+    auto broker = _members_table.local().get_broker(target);
+    if (!broker) {
+        co_return make_error_code(errc::node_does_not_exists);
+    }
+
+    if (!_node_connection_cache.contains(target)) {
+        auto cert = co_await maybe_build_reloadable_certificate_credentials(
```

Comment:
nit: there is this additional check in cluster_utils that checks if rpc_address of a node_id changed and invalidates the connection, that seems like a legit usecase. perhaps incorporate that? 


```

ss::future<> maybe_create_tcp_client(
  rpc::connection_cache& cache,
  model::node_id node,
  net::unresolved_address rpc_address,
  config::tls_config tls_config) {
    auto f = ss::now();
    if (cache.contains(node)) {
        // client is already there, check if configuration changed
        if (cache.get(node)->server_address() == rpc_address) {
            // If configuration did not changed, do nothing
            return f;
        }
        // configuration changed, first remove the client
        f = cache.remove(node);
    }
    // there is no client in cache, create new
```

Also as a reader, it felt like the connection_cache management code here (especially around tls certs), seemed a bit out of place but other places like members_manager are doing the same so its ok.

Reply:
Good point on detecting rpc address changes. A node could re-start with a different address.

> Also as a reader, it felt like the connection_cache management code here (especially around tls certs), seemed a bit out of place but other places like members_manager are doing the same so its ok.

I agree. It feels too low level somehow. I think some of the stuff in cluster utils should be folded into the connection cache to prevent this duplication.

Reply:
Went back to the code and I remembered why I duplicated this logic. It was in order to disable the metrics for the connection. Currently, the labels used for the transport metrics are the target host and port. Since this new connection is made to the RPC server, the host and port the labels clash. This causes double registration errors. I'd like to keep this PR low on refactoring and address that separately. Once that works fine, I can use `maybe_create_tcp_client` from cluster utils here.

Reply:
I fixed the metrics issue and switched to using `maybe_create_tcp_client`.